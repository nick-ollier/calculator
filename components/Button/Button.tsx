import { useState } from 'react';
import classnames from 'classnames';
import DOMPurify from 'isomorphic-dompurify';

const c = {
    number: 'bg-gray-400 text-gray-100',
    numberActive: 'bg-gray-500',
    defaults: 'bg-gray-200 text-gray-500',
    defaultsActive: 'bg-gray-300',
    operator: 'bg-orange-200 text-gray-100 dark:bg-orange-100',
    operatorActive: 'bg-white text-orange-200 dark:text-orange-100',
    scientific: 'bg-gray-500 text-gray-100',
    scientificActive: 'bg-gray-600',
    portrait: 'min-w-btn-p min-h-btn-p',
    landscape: 'min-h-btn-l min-w-btn-l',
    span: 'text-left px-6'
};

export interface ButtonProps {
    span?: number;
    variant?: 'default' | 'number' | 'operator' | 'scientific';
    active: boolean;
    label: string;
    accessibleLabel: string;
    value: string;
    portrait: boolean;
    onClick: () => void;
}

const Button: React.FC<ButtonProps> = ({
    span = 1,
    variant = 'default',
    active,
    label,
    accessibleLabel,
    value,
    portrait,
    onClick
}) => {
    const [displayClick, setDisplayClick] = useState(false);

    const handleClick = () => {
        setDisplayClick(true);
        onClick();

        setTimeout(() => {
            setDisplayClick(false);
        }, 300);
    };

    return (
        <button
            type='button'
            value={value}
            aria-label={accessibleLabel}
            onClick={handleClick}
            className={classnames('rounded-full', `col-span-${span}`, {
                [c.number]: variant === 'number',
                [c.numberActive]: variant === 'number' && displayClick,
                [c.defaults]: variant === 'default',
                [c.defaultsActive]: variant === 'default' && displayClick,
                [c.operator]: variant === 'operator' && !active,
                [c.operatorActive]: variant === 'operator' && active,
                [c.scientific]: variant === 'scientific',
                [c.scientificActive]: variant === 'scientific' && displayClick,
                [c.span]: span > 1,
                [c.portrait]: portrait,
                [c.landscape]: !portrait
            })}
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{
                __html: DOMPurify.sanitize(label)
            }}
        />
    );
};

export default Button;
