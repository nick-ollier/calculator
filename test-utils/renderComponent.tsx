import CalculatorProvider from '@context/CalculatorContext';
import { render } from '@testing-library/react';

const renderWrapper = (ui: React.ReactElement) => {
    const WithProviders = ({ children }) => {
        return <CalculatorProvider>{children}</CalculatorProvider>;
    };

    return render(ui, { wrapper: WithProviders });
};

export * from '@testing-library/react';
export { renderWrapper as render };
