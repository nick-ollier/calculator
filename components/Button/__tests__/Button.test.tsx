import * as React from 'react';
import { render, screen } from '@testUtils/renderComponent';
import userEvent from '@testing-library/user-event';
import Button, { ButtonProps } from '@components/Button/Button';

const mockOnClick = jest.fn();

const renderComponent = (props?: Partial<ButtonProps>) => {
    const utils = render(
        <Button
            active={false}
            label='button'
            accessibleLabel='button'
            value='button'
            portrait
            onClick={() => null}
            {...props}
        />
    );

    return {
        ...utils,
        button: screen.queryByRole('button', { name: /button/i })
    };
};

describe('component: button', () => {
    describe('portrait', () => {
        describe('default', () => {
            it('renders', () => {
                const { container } = renderComponent();
                expect(container);
            });

            it('renders with correct classes', () => {
                const { button } = renderComponent();
                expect(button).toHaveClass('bg-gray-200 text-gray-500');
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent();
                expect(asFragment()).toMatchSnapshot();
            });

            it("shouldn't fire callback on click when a callback isn't provided", () => {
                const { button } = renderComponent();

                userEvent.click(button);
                expect(mockOnClick).not.toHaveBeenCalled();
            });

            it('fires callback on click', () => {
                const { button } = renderComponent({ onClick: mockOnClick });

                userEvent.click(button);
                expect(mockOnClick).toHaveBeenCalled();
            });

            it('sanitizes label on render', () => {
                renderComponent({
                    label: "<script>console.log('not today!')</script>Hello!",
                    accessibleLabel:
                        "<script>console.log('not today!')</script>Hello!"
                });

                expect(
                    screen.getByRole('button', { name: /hello!/i })
                ).toBeInTheDocument();
            });

            it('updates button state on click, and again after timeout', () => {
                const setState = jest.fn();

                const useStateMock: any = (initState: any) => [
                    initState,
                    setState
                ];

                // window.setTimeout = jest.fn();

                jest.useFakeTimers();
                jest.spyOn(React, 'useState').mockImplementation(useStateMock);

                const { button } = renderComponent();

                userEvent.click(button);

                expect(setState).toHaveBeenCalledTimes(1);
                expect(setState).toHaveBeenCalledWith(true);
                jest.runAllTimers();
                expect(setState).toHaveBeenCalledTimes(2);
                expect(setState).toHaveBeenCalledWith(false);

                jest.clearAllMocks();
            });
        });

        describe('variants', () => {
            it('number - renders', () => {
                const { container } = renderComponent({ variant: 'number' });
                expect(container);
            });

            it('number - renders with correct classes', () => {
                const { button } = renderComponent({ variant: 'number' });
                expect(button).toHaveClass('bg-gray-400 text-gray-100');
            });

            it('number - matches snapshot', () => {
                const { asFragment } = renderComponent({ variant: 'number' });
                expect(asFragment()).toMatchSnapshot();
            });

            it('operator - renders', () => {
                const { container } = renderComponent({ variant: 'operator' });
                expect(container);
            });

            it('operator - renders with correct classes', () => {
                const { button } = renderComponent({ variant: 'operator' });
                expect(button).toHaveClass(
                    'bg-orange-200 text-gray-100 dark:bg-orange-100'
                );
            });

            it('operator - matches snapshot', () => {
                const { asFragment } = renderComponent({ variant: 'operator' });
                expect(asFragment()).toMatchSnapshot();
            });

            it('scientific - renders', () => {
                const { container } = renderComponent({
                    variant: 'scientific'
                });
                expect(container);
            });

            it('scientific - renders with correct classes', () => {
                const { button } = renderComponent({ variant: 'scientific' });
                expect(button).toHaveClass('bg-gray-500 text-gray-100');
            });

            it('scientific - matches snapshot', () => {
                const { asFragment } = renderComponent({
                    variant: 'scientific'
                });
                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe('with props', () => {
            it('renders with correct span class', () => {
                const { button } = renderComponent({ span: 2 });
                expect(button).toHaveClass('col-span-2');
            });
        });
    });

    describe('landscape', () => {
        describe('default', () => {
            it('renders', () => {
                const { container } = renderComponent({ portrait: false });
                expect(container);
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent({ portrait: false });
                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
