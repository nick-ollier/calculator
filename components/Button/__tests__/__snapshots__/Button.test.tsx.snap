// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: button landscape default matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="button"
    class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-h-btn-l min-w-btn-l"
    type="button"
    value="button"
  >
    button
  </button>
</DocumentFragment>
`;

exports[`component: button portrait default matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="button"
    class="rounded-full col-span-1 bg-gray-200 text-gray-500 min-w-btn-p min-h-btn-p"
    type="button"
    value="button"
  >
    button
  </button>
</DocumentFragment>
`;

exports[`component: button portrait variants number - matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="button"
    class="rounded-full col-span-1 bg-gray-400 text-gray-100 min-w-btn-p min-h-btn-p"
    type="button"
    value="button"
  >
    button
  </button>
</DocumentFragment>
`;

exports[`component: button portrait variants operator - matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="button"
    class="rounded-full col-span-1 bg-orange-200 text-gray-100 dark:bg-orange-100 min-w-btn-p min-h-btn-p"
    type="button"
    value="button"
  >
    button
  </button>
</DocumentFragment>
`;

exports[`component: button portrait variants scientific - matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="button"
    class="rounded-full col-span-1 bg-gray-500 text-gray-100 min-w-btn-p min-h-btn-p"
    type="button"
    value="button"
  >
    button
  </button>
</DocumentFragment>
`;
